src/zope/paste/tests.py

```diff
@@ -124,14 +124,27 @@ def test_basic_serve():
 def test_serving_test_app():
     """Test serving a real app.
 
+    Setup logging, so we get the message ``Serving on ...`` into our doctest
+    output:
+
+    >>> import logging
+    >>> logger = logging.getLogger('waitress')
+    >>> logger.addHandler(logging.StreamHandler(sys.stdout))
+    >>> logger.setLevel(logging.INFO)
+
+    Start the app:
+
+    >>> import os
+    >>> import os.path
     >>> import threading
     >>> from zope.paste import serve
-    >>> server = threading.Thread(target=serve.serve, args=(['app.ini'],))
+    >>> app_ini = os.path.join(os.getcwd(), 'app.ini')
+    >>> server = threading.Thread(target=serve.serve, args=([app_ini],))
     >>> server.start()
     >>> import time; time.sleep(5)  # doctest: +ELLIPSIS
-    Serving on http://localhost:...
+    Serving on http://...
 
-    >>> print(urlopen('http://localhost:{}/'.format(PORT)).read().decode())
+    >>> print(urlopen('http://127.0.0.1:{}/'.format(PORT)).read().decode())
```

Comment:
Oh, that is interesting.

I once had problems where connections to localhost:NNNN failed but 127.0.0.1:NNNN succeeded, because my /etc/hosts had localhost resolve to ::1, but the service on port NNNN was listening on IPv4 only.

Reply:
When using `localhost` `waitress >= 2` opens both on IPv4 and IPv6 but this is not backwards compatible with `waitress 1.x` which is still needed for Python 2.7 as `waitress >= 2` dropped Python 2 support.